### PR TITLE
Make alert dismissable

### LINKED ISSUE #4236: Ability to make the alerts dismiss able

### PR DESCRIPTION
**Pull request by hishamco**
Fix #4236

Alert

### CODE REVIEW

File: src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs [hunk 1]
```diff
@@ -104,10 +104,15 @@ public IHtmlContent Message(dynamic Shape)
             TagBuilder tagBuilder = OrchardCore.DisplayManagement.Shapes.Shape.GetTagBuilder(Shape, "div");
             string type = Shape.Type.ToString().ToLowerInvariant();
             IHtmlContent message = Shape.Message;
-            tagBuilder.AddCssClass("message");
-            tagBuilder.AddCssClass("message-" + type);
+            tagBuilder.AddCssClass("alert");
+            tagBuilder.AddCssClass("alert-" + type);
+            tagBuilder.AddCssClass("alert-dismissible fade show");
             tagBuilder.Attributes["role"] = "alert";
```
Comment (deanmarcussen) [inline src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs]: I like the idea, but not the implementation sorry

@agriffard wanted to do exactly the same thing on https://github.com/OrchardCMS/OrchardCore/pull/4538 but decided to implement our own css to do so and keep the `message` class

See https://github.com/OrchardCMS/OrchardCore/pull/4579 for how it was done. 

Suggest you extend that in same way, as @ns8482e is right, this shape is designed to be used by themes that are not implementing bootstrap

Comment (hishamco) [reply src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs]: This is how it was before, if you like the shape style we need to extract the entire markup

Comment (deanmarcussen) [reply src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs]: @hishamco if you have a look in `TheAdmin.scss` file you'll find the messages section where bootstrap has already been extended.

```
// Messages
// ------------------------------
.message {
    @extend .alert;
}

.message-success {
    @extend .alert-success;
}

.message-information {
    @extend .alert-info;
}

.message-warning {
    @extend .alert-warning;
}

.message-error {
    @extend .alert-danger;
}
```

you should just be able to add `@extend .alert-dismissable` and the fade / show mixins to the sass without changing any of the Shape classes, or having to extract the entire markup

Comment (hishamco) [reply src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs]: I got what you mean, let me try

Comment (hishamco) [reply src/OrchardCore/OrchardCore.DisplayManagement/Shapes/CoreShapes.cs]: I added `@extend .alert-dismissable` but the close button doesn't work as expected

**Comment by hishamco** [comment]
@deanmarcussen could you please help me on this if you have a time

**Comment by deanmarcussen** [comment]
Hmm, I guess it doesn't work because the bootstrap javascript requires the correct class name.
You could bring that javascript across, but that sounds hard, and harder to keep maintained.

Why don't you just override the `Message` shape in `TheAdmin` theme, with a razor view. eg.
`Message.cshtml` and build the tag builder as required there? 
Something like...
```
@model dynamic
@{
   tagBuilder = Tag(Model, "div");
   ... (all the variables from the Shape will be available on Model)
}

@tagBuilder

**Comment by ns8482e** [comment]
@hishamco  @deanmarcussen  If you are implementiong for TheAdmin theme then instead of changing core shape you could create an alternate in TheAdmin theme that use "altert" class instead of "message".

**Comment by hishamco** [comment]
> f you are implementing for TheAdmin

Yes

> you could create an alternate in TheAdmin theme and use "alert" class instead of "message"

How? I'm not familiar with alternate

**Comment by ns8482e** [comment]
Using `IShapeTableProvider` you can add alternates to any shape using `OnDisplaying` event.

For example
```
 builder.Describe("Message").OnDisplaying(context =>
                {
                    dynamic msgShape = context.Shape;
                    msgShape.Metadata.Alternates.Add("Message__admin");
                });
```
and create `Message-admin.cshtml` in admin theme

**Comment by hishamco** [comment]
You remind me with an old example that I stuck on it before :smile:

thanks so much

**Comment by ns8482e** [comment]
@hishamco  you are welcome

**Comment by deanmarcussen** [comment]
cool @ns8482e and @hishamco.

For infos in this case you wouldn't need to create an alternate with an `IShapeTableProvider`.

You can simply override the shape by using `Message.cshtml` in TheAdmin theme.

This will only affect the admin theme, as the front end themes will always use a different shape table (with a different order of precedence.)

The priority of shape precedence makes this possibly.
It goes something like
...modules 
TheAdmin (Theme)
SuperTheme (i.e. startup project, which can override a Theme shape and take precedence.

**Comment by hishamco** [comment]
@Skrypt if someone can continue on this no problem. We 're facing an Internet cable issue in my country, but I will struggle

**Comment by hishamco** [comment]
@Skrypt if no one will update this I will try to have time to finish this one

**Comment by Skrypt** [comment]
Moved to https://github.com/OrchardCMS/OrchardCore/pull/5587

**Pull request closed without merging**